File: docs/algorithms/code/algorithms/generateMatrix.ts
// 本质上是要区分好区间
function generateMatrix(n: number): number[][] {
  const res: number[][] = new Array(n).fill(0).map(() => new Array(n).fill(0));
  let num = 1;

  let left = 0, right = n - 1, top = 0, bottom = n - 1;
  while (left <= right && top <= bottom) {
    for (let col = left; col <= right; col++) {
      res[top][col] = num++;
    }
    if (top++ > bottom) {
      break;
    }
    for (let row = top; row <= bottom; row++) {
      res[row][right] = num++;
    }
    if (right-- < left) {
      break;
    }
    for (let col = right; col >= left; col--) {
      res[bottom][col] = num++;
    }
    if (bottom-- < top) {
      break;
    }

    for (let row = bottom; row >= top; row--) {
      res[row][left] = num++;
    }
    if (left++ > right) {
      break;
    }
  }

  return res;
};

File: docs/algorithms/code/algorithms/generateMatrix.go
// 思路：使用四个边界指针模拟顺时针螺旋填充过程，每填充完一条边就收缩对应边界，直到填满整个矩阵
// 复杂度分析：时间复杂度 O(n^2), 空间复杂度 O(n^2)
func generateMatrix(n int) [][]int {
	arr := make([][]int, n)

	for i := 0; i < n; i++ {
		arr[i] = make([]int, n)
	}

	count, max := 1, n*n

	left, right, top, bottom := 0, n-1, 0, n-1

	for count <= max {
		for i := left; i <= right; i++ {
			arr[top][i] = count
			count++
		}
		top++

		for i := top; i <= bottom; i++ {
			arr[i][right] = count
			count++
		}
		right--

		for i := right; i >= left; i-- {
			arr[bottom][i] = count
			count++
		}
		bottom--

		for i := bottom; i >= top; i-- {
			arr[i][left] = count
			count++
		}
		left++
	}

	return arr
}